### PR TITLE
LMDB Integration Test

### PR DESCRIPTION
**Pull request by scorbajio**
This change tests and benchmarks integrating LMDB for use as the primary state DB. Also see #2565.

### CODE REVIEW

**Comment by holgerd77** [comment]
Hey there, cool that this got a start! 🤩 Just to point this out explicitly so things are not mixed up: in the first round this only about *testing* *if* LMDB is a viable option for the client. If we want to switch is far from being certain and would be a separate discussion after we have the results from this PR.

Will give this a "do not merge" label to make this more clear and to avoid that this gets accidentally merged in.

File: packages/client/src/blockchain/chain.ts [hunk 1]
```diff
@@ -2,14 +2,108 @@ import { Block, BlockHeader } from '@ethereumjs/block'
...
+  shallowCopy(): DB<TKey, TValue> {
+    // @ts-ignore
+    return
+  }
+
+  async open() {}
+
+  async close() {}
+}
+
```
Comment (holgerd77) [inline packages/client/src/blockchain/chain.ts:106]: Not sure if this is meant to be permanently here, but this should be in a separate file `lmdb.ts`.

Comment (holgerd77) [reply packages/client/src/blockchain/chain.ts:106]: (just to note, if we go this way).

**Comment by holgerd77** [comment]
So for me `npm run client:start` is not yet starting on this branch (installed `lmdb` globally for now).

**Comment by scorbajio** [comment]
> So for me `npm run client:start` is not yet starting on this branch (installed `lmdb` globally for now).

For me, if it hangs on initializing vm execution, I give it a few minutes (about 5) and it finishes generating genesis and continues with fetching and executing blocks as expected. Is that the issue or are you facing a different one?

**Comment by scorbajio** [comment]
I only went to the 200,000th block since on the LMDB branch it was already at 4 hours. I have some results to share:

on lmdb-integration (LMDB as stateDB): ~4 hours

```
[09-05|14:14:23] INFO Data directory: /Users/indigoalpha/Library/Ethereum/ethereumjs
...
[09-05|18:15:13] INFO Executed blocks count=100 first=199900 hash=0x8cf4…06a0 td=3522739844945208155 hardfork=chainstart last=200000 hash=0x13ce…f693 txs=77 
```

on master (LevelDB as stateDB): ~2 hours 10 minutes

```
[09-05|18:29:37] INFO Data directory: /Users/indigoalpha/Library/Ethereum/ethereumjs
...
[09-05|20:41:28] INFO Executed blocks count=100 first=199900 hash=0x8cf4…06a0 td=1752020526493328466 hardfork=chainstart last=200000 hash=0x13ce…f693 txs=77 (edited)
```

I'll do a quick test to benchmark how read/write speeds compare between LMDB and LevelDB do make sure the results make sense.

**Comment by scorbajio** [comment]
LevelDB:

1k-3-32-ran x 2 ops/sec @ 409ms/op
1k-5-32-ran x 2 ops/sec @ 412ms/op
1k-9-32-ran x 2 ops/sec @ 418ms/op
1k-1k-32-ran x 2 ops/sec @ 488ms/op
1k-1k-32-mir x 2 ops/sec @ 450ms/op
Checkpointing: 100 iterations x 2,273 ops/sec @ 439μs/op ± 9.91% (min: 156μs, max: 2ms)
Checkpointing: 500 iterations x 2,359 ops/sec @ 423μs/op ± 3.42% (min: 256μs, max: 1ms)
Checkpointing: 1000 iterations x 2,340 ops/sec @ 427μs/op ± 2.33% (min: 277μs, max: 1ms)
Checkpointing: 5000 iterations x 2,018 ops/sec @ 495μs/op ± 1.04% (min: 290μs, max: 2ms)
RAM: rss=384.5mb heap=239.0mb used=206.2mb ext=61.2mb arr=59.8mb

---

LMDB

1k-3-32-ran x 0 ops/sec @ 2793ms/op
1k-5-32-ran x 0 ops/sec @ 3037ms/op
1k-9-32-ran x 0 ops/sec @ 3181ms/op
1k-1k-32-ran x 0 ops/sec @ 3148ms/op
1k-1k-32-mir x 0 ops/sec @ 3252ms/op
Checkpointing: 100 iterations x 369 ops/sec @ 2ms/op ± 8.44% (min: 1ms, max: 9ms)
Checkpointing: 500 iterations x 344 ops/sec @ 2ms/op ± 2.83% (min: 1ms, max: 6ms)
Checkpointing: 1000 iterations x 308 ops/sec @ 3ms/op ± 1.81% (min: 1ms, max: 19ms)
Checkpointing: 5000 iterations x 287 ops/sec @ 3ms/op
RAM: rss=351.6mb heap=201.6mb used=178.1mb ext=5.8mb arr=11.1mb

**Comment by scorbajio** [comment]
With the test results from this PR, it can be concluded that switching to LMDB will not provide a performance gain. I have created #3016 as a followup for adding LMDB to the trie package benchmarks and fixing the race condition from this PR that was discovered at commit hash `ce1a289948f05f0b944b7261f13f6e955dd1c441` and fixed at commit hash `00426fc08ad23f2f74c5c6f8746f554e282cddf1`.

**Comment by holgerd77** [comment]
Thanks a lot for investigating this! 🙏

**Comment by holgerd77** [comment]
Will close the LMDB test PRs for now. This is open for quite some time and did not lead to anything usable up till now. I additionally stumbled upon some more detailed benchmarks (cannot re-find the reference unfortunately) stating that LMDB will not bring performance gains for larget datasets.

RocksDB (Facebook) might still be an option to explore. We'll see.

**Pull request closed without merging**